spec.html

```diff
@@ -149,6 +149,37 @@ <h1>
         1. Return _iteratorRecord_.
       </emu-alg>
     </emu-clause>
+
+    <emu-clause id="sec-getiteratorflattenable" type="abstract operation">
+      <h1>
+        GetIteratorFlattenable (
+          _obj_: an ECMAScript language value,
+          optional _hint_: ~sync~ or ~async~,
+        ): either a normal completion containing an Iterator Record or a throw completion
+      </h1>
+      <dl class="header">
+      </dl>
+      <emu-alg>
+        1. If _hint_ is not present, set _hint_ to ~sync~.
+        1. If _iterator_ is either *undefined* or *null*, throw a *TypeError* exception.
+        1. Let _method_ be *undefined*.
+        1. If _hint_ is ~async~, then
+          1. Set _method_ to ? GetV(_obj_, @@asyncIterator).
+        1. If IsCallable(_method_) is *false*, then
+          1. Set _method_ to ? GetV(_obj_, @@iterator).
+        1. If IsCallable(_method_) is *false*, then
+          1. Let _iterator_ be _obj_.
+        1. Else,
+          1. Let _iterator_ be ? Call(_method_, _obj_).
+        1. If _iterator_ is not an Object, throw a *TypeError* exception.
+        1. Let _nextMethod_ be ? Get(_iterator_, *"next"*).
+        1. If IsCallable(_nextMethod_) is *false*, throw a *TypeError* exception.
+        1. Let _iteratorRecord_ be the Iterator Record { [[Iterator]]: _iterator_, [[NextMethod]]: _nextMethod_, [[Done]]: *false* }.
+        1. If _hint_ is ~async~, then
+          1. Return CreateAsyncFromSyncIterator(_iteratorRecord_).
+        1. Return _iteratorRecord_.
```

Comment:
Using `GetIterator`:
```suggestion
        1. If _hint_ is not present, set _hint_ to ~sync~.
        1. If _iterator_ is either *undefined* or *null*, throw a *TypeError* exception.
        1. If _hint_ is ~async~, then
          1. Let _usingAsync_ be ? GetMethod(_obj_, @@asyncIterator).
          1. If _usingAsync_ is not *undefined*, then
            1. Return ? GetIterator(_obj_, ~async~, _usingAsync_).
        1. Let _usingSync_ be ? GetMethod(_obj_, @@iterator).
        1. If _usingSync_ is not *undefined*, then
          1. Let _syncIteratorRecord_ be ? GetIterator(_obj_, ~sync~, _usingSync_)
          1. If _hint_ is ~async~, then
            1. Return CreateAsyncFromSyncIterator(_syncIteratorRecord_)
          1. Else,
            1. Return _syncIteratorRecord_
        1. Let _nextMethod_ be ? GetMethod(_iterator_, *"next"*).
        1. Return the Iterator Record { [[Iterator]]: _iterator_, [[NextMethod]]: _nextMethod_, [[Done]]: *false* }.
```
I would in fact refactor `GetIterator` to separate the two-parameter and the three-parameter versions. Instead:
* `GetIterator(obj, iteratorMethod): IteratorRecord { return CreateIteratorRecord(Call(obj, iteratorMethod)) }`
* `CreateIteratorRecord(iterator): IteratorRecord { return {[[Iterator]]: iterator, [[NextMethod]]: GetMethod(iterator, "next"), …} }`
* `IterableToIterator(obj, hint): IteratorRecord { GetMethod … GetIterator … CreateAsyncFromSyncIterator }` - what `GetIterator` without a method currently does

Reply:
Such refactoring would also benefit https://github.com/tc39/ecma262/issues/746 - apparently engines don't implement the extra object check in `GetIteratorDirect`, they happily access `.next` on primitive values as well if `[Symbol.iterator]` returns one.